src/diff/mod.rs

```diff
@@ -102,65 +97,81 @@ impl Diff {
-                                (File(new_file), File(old_file)) => {
-                                    if old_file.size != new_file.size
-                                        || old_file.checksum() != new_file.checksum()
-                                    {
-                                        diff.add_modified_file(
-                                            new_file,
-                                            &RefCell::borrow(parent_path),
-                                        );
-                                    }
-                                    old_entry_opt = old_iter.next();
-                                    new_entry_opt = new_iter.next();
-                                }
-                                (File(new_file), SubDirectory(old_dir)) => {
-                                    diff.add_created_file(new_file, &RefCell::borrow(parent_path));
-                                    diff.add_deleted_files(old_entry, parent_path)?;
-                                    old_entry_opt = old_iter.next();
-                                    new_entry_opt = new_iter.next();
-                                }
-                                (SubDirectory(new_dir), File(old_file)) => {
-                                    diff.add_created_files(new_entry, parent_path)?;
-                                    diff.add_deleted_file(old_file, &RefCell::borrow(parent_path));
-                                    old_entry_opt = old_iter.next();
-                                    new_entry_opt = new_iter.next();
-                                }
-                                (SubDirectory(new_dir), SubDirectory(old_dir)) => {
-                                    parent_path.borrow_mut().push(new_dir.name.clone());
-                                    Diff::collect_diff(
-                                        old_dir.deref(),
-                                        new_dir.deref(),
-                                        parent_path,
-                                        diff,
-                                    )?;
-                                    parent_path.borrow_mut().pop();
-                                    old_entry_opt = old_iter.next();
-                                    new_entry_opt = new_iter.next();
-                                }
-                                // need to skip Repos
-                                (Repo, Repo) => {
-                                    old_entry_opt = old_iter.next();
-                                    new_entry_opt = new_iter.next();
-                                }
-                                (Repo, _) => {
-                                    old_entry_opt = old_iter.next();
-                                }
-                                (_, Repo) => {
-                                    new_entry_opt = new_iter.next();
+                        Ordering::Equal => match (new_entry, old_entry) {
```

Comment:
These triple-nested match shenanigans become quite hard to dissect as a reader, maybe the big body can be broken apart.